File: problems/1031_maximum_sum_of_two_non_overlapping_subarrays/Solution.java
class Solution {
    public int maxSumTwoNoOverlap(int[] nums, int firstLen, int secondLen) {
        return Math.max(solve(nums, firstLen, secondLen), solve(nums, secondLen, firstLen));
    }
    private int solve(int[] nums, int L, int M) {
        int n = nums.length;
        int[] prefix = new int[n + 1];
        for (int i = 0; i < n; i++) prefix[i+1] = prefix[i] + nums[i];
        int maxL = 0, ans = 0;
        for (int i = L + M; i <= n; i++) {
            maxL = Math.max(maxL, prefix[i - M] - prefix[i - M - L]);
            ans = Math.max(ans, maxL + prefix[i] - prefix[i - M]);
        }
        return ans;
    }
}

File: problems/1031_maximum_sum_of_two_non_overlapping_subarrays/solution.py
class Solution:
    def maxSumTwoNoOverlap(self, nums: list[int], firstLen: int, secondLen: int) -> int:
        def solve(L: int, M: int) -> int:
            prefix = [0] * (len(nums) + 1)
            for i in range(len(nums)):
                prefix[i+1] = prefix[i] + nums[i]
            maxL = 0
            ans = 0
            for i in range(L + M, len(nums) + 1):
                maxL = max(maxL, prefix[i - M] - prefix[i - M - L])
                ans = max(ans, maxL + prefix[i] - prefix[i - M])
            return ans
        return max(solve(firstLen, secondLen), solve(secondLen, firstLen))
